service/storage_proxy.cc

```diff
@@ -5309,6 +5310,68 @@ future<bool> storage_proxy::check_hint_queue_sync_point(utils::UUID sync_point)
     });
 }
 
+future<> storage_proxy::wait_for_hints_to_be_replayed(utils::UUID operation_id, std::vector<gms::inet_address> source_endpoints, std::vector<gms::inet_address> target_endpoints) {
+    auto& db = _db.local();
+    const auto timeout_in_ms = db.get_config().wait_for_hint_replay_before_repair_in_ms();
+    if (timeout_in_ms <= 0) {
+        slogger.debug("wait_for_hint_replay_before_repair_in_ms is zero, so I'm not waiting for other nodes to replay their hints");
+        co_return;
+    }
+
+    const auto deadline = lowres_clock::now() + std::chrono::milliseconds(timeout_in_ms);
+
+    slogger.debug("Coordinating a request to wait until all hints are sent on {} nodes", target_endpoints.size());
+
+    co_await parallel_for_each(source_endpoints, [this, operation_id,  &target_endpoints_ = target_endpoints, deadline_ = deadline] (gms::inet_address addr) -> future<> {
+        const auto deadline = deadline_;
+        const auto& target_endpoints = target_endpoints_;
+
+        const std::chrono::seconds wait_duration{1};
+
+        // Step 1: Create a hint queue sync point
+        try {
+            co_await _messaging.send_hint_sync_point_create({ addr, 0 }, deadline, operation_id, target_endpoints, deadline);
+        } catch (...) {
+            slogger.debug("Failed to create a sync point for {}. I won't wait for hints to be replayed on this node");
+            throw;
+        }
+
+        // Step 2: Wait until hints are replayed
+        // Sleep 1 second between checks
+        bool reached_sync_point = false;
+        while (!reached_sync_point && lowres_clock::now() < deadline) {
+            slogger.debug("Waiting for all hints from endpoint {} to be sent out; remaining time: {}s, targets: {}",
+                    addr, std::chrono::duration_cast<std::chrono::seconds>(deadline - lowres_clock::now()).count(), target_endpoints);
+            try {
+                reached_sync_point = co_await _messaging.send_hint_sync_point_check({ addr, 0 }, deadline, operation_id);
+                if (reached_sync_point) {
+                    break;
+                }
+            } catch (rpc::timeout_error&) {
+                slogger.debug("Caught a timeout error while sending hint queue query at {}", addr);
+                // RPC timeout should be set to the deadline, so we can assume here that the deadline is reached
+                break;
+            } catch (rpc::closed_error&) {
+                slogger.debug("Caught a connection closed error while sending hint queue query at {}", addr);
```

Comment:
Shouldn't we exit? If the node is down, we'll wait a lot. Since we allow hints to accumulate for hours, replay could take hours too. Especially if we tame replay to not overwhelm the receiver.

(btw, maybe the parameter should not be in units of milliseonds)

Reply:
I assumed (maybe wrongly) that a closed connection can happen as a temporary error - I wanted the repair coordinator to be resilient to such events. A later commit introduces a check which exits this loop if the remote endpoint is detected as not alive (`storage_proxy: stop waiting for hints replay when node goes down`).

I'll change it so that the loop stops on this error, but if you know if there are some temporary errors which can be ignored, I can add special handling for them so that they are ignored.

Reply:
Changed units: now the timeout parameter is not `wait_for_hint_replay_before_repair_in_ms`, it's `wait_for_hint_replay_before_repair_in_seconds`